scripts/tprs-data-extract/1211-three-years-or-more.js

```diff
@@ -0,0 +1,116 @@
+                     from form as f
+                     where status = 'APPROVED'
+                         ${onlyTprs ? `AND form_response -> 'openConditions' -> 'tprs' ->> 'tprsSelected' = 'Yes'` : ''}
+                       AND (approval_date >= COALESCE($1::date, (SELECT MIN(approval_date) FROM public.form)))
+                       AND (approval_date <= COALESCE($2::date, CURRENT_DATE))
+                       AND prison_id NOT IN
+                           ('AGI', 'DWI', 'DHI', 'ESI', 'EWI', 'BZI', 'FHI', 'LNI', 'SDI', 'STI', 'NHI', 'PFI')
+                     )
+      select *
+from prisoners_table
+where cat = 'C'
+`
+
+/**
+ * @typedef {object} Booking
+ * @property {string} offender_no
+ * @property {number} booking_id
+ * @property {number} sequence_no
+ * @property {string} prison_id
+ * @property {string} cat_type
+ * @property {string} cat
+ */
+
+/**
+ * @param {boolean} onlyTprs
+ * @param {string} startDate
+ * @param {string} endDate
+ * @returns {Promise<Booking[][]>}
+ */
+const getCatCPrisoners = async ({ onlyTprs, startDate, endDate }) => {
+  /**
+   * @type {QueryArrayResult<Booking[]>}
+   */
+  const result = await db.query(query({ onlyTprs }), [startDate, endDate])
+  return result.rows
+}
+const createBookingsMap = ({ bookings }) => {
+  const bookingMap = new Map()
+
+  bookings.forEach(booking => {
+    bookingMap.set(booking.booking_id.toString(), booking)
+  })
+
+  return bookingMap
+}
+/**
+ * @param {Booking[]} bookings
+ * @returns {Promise<Booking[][]>}
+ */
+const getPrisonerSearchData = async ({ username, bookingIds }) => {
+  const prisonerSearchClient = prisonerSearchClientBuilder({ user: { username } })
+  return prisonerSearchClient.getPrisonersByBookingIds([...bookingIds])
+}
+
+const updateBookingMapWithSearchResult = ({ bookingMap, prisonerSearchData }) => {
+  prisonerSearchData.forEach(result => {
+    // console.log('111', result)
+    // ensure consistent key format
+    const bookingIdStr = result.bookingId.toString()
+    const existingBooking = bookingMap.get(bookingIdStr)
+    // console.log('222', existingBooking)
+    const updatedBooking = {
+      ...existingBooking,
+      ...result,
+      releaseDateObject: new Date(result.releaseDate),
+    }
+    // console.log('updatedBooking', updatedBooking)
+    bookingMap.set(bookingIdStr, updatedBooking)
+  })
+
+  return bookingMap
+}
+
+module.exports = {
+  getCatCPrisonersWith3OrMoreYearsRemaining: async ({ username, onlyTprs, startDate, endDate }) => {
+    const bookings = await getCatCPrisoners({ onlyTprs, startDate, endDate })
+
+    const bookingMap = createBookingsMap({ bookings })
+    const bookingIds = bookingMap.keys()
+
+    const prisonerSearchData = await getPrisonerSearchData({ username, bookingIds })
+    const updatedBookingMap = updateBookingMapWithSearchResult({ bookingMap, prisonerSearchData })
+    const threeYearsFromNow = moment().add(3, 'years')
+
+    return [...updatedBookingMap.values()]
```

Comment:
Do you need to splay this into an array? does updatedBookingMap.values().filter not work?